v2/index/index.go

```diff
@@ -40,7 +43,7 @@ type (
 		Codec() multicodec.Code
 
 		// Marshal encodes the index in serial form.
-		Marshal(w io.Writer) error
+		Marshal(w io.Writer) (uint64, error)
```

Comment:
Is anything using this index package directly where this signature change would break them?

Reply:
I think the answer is we don't know. Would be great if we don't break the API if we don't have to and use standard interfaces, specially when the signature matches exactly

Reply:
This is the trap IPFS fell into. the answer can't be 'because we don't know we are unwilling to deprecate any API'. I think we stick with the smaller API that we want, rather than keeping all previous versions in 'deprecated'

Reply:
I need to learn more about what happened in IPFS. In this specific case I do think it makes sense to reuse an existing interface used already elsewhere. From API design perspective at the very least. 

This is not a blocker for the purposes of this PR review from my perspective.